enumerables.rb

```diff
@@ -0,0 +1,203 @@
+module Enumerable
+  # my_each
+
+  def my_each
+    if block_given?
+      if is_a?(Range)
+        arr = to_a
+        arr.each do |value|
+          yield(value)
+        end
+      else
+        each do |value|
+          yield(value)
+        end
+      end
+      self
+    else
+      to_enum :my_each
+    end
+  end
+
+  # my_each_with_index
+  def my_each_with_index
+    if block_given?
+      arr = if is_a?(Range)
+              to_a
+            else
+              self
+            end
+      (0..arr.length - 1).each do |index|
+        yield(arr[index], index)
+      end
+      self
+    else
+      to_enum :my_each_with_index
+    end
+  end
```

Comment:
Your `my_each_with_index` returns the following error for automated tests:

```
1) enumerables #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self
     Failure/Error: expect(my_each_with_index_output).to eq(each_with_index_output)
     
       expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
            got: "Num: , idx: 0\nNum: , idx: 1\nNum: , idx: 2\nNum: , idx: 3\nNum: , idx: 4\n"
     
       (compared using ==)
     
       Diff:
       @@ -1,6 +1,6 @@
       -Num: [:a, 1], idx: 0
       -Num: [:b, 2], idx: 1
       -Num: [:c, 3], idx: 2
       -Num: [:d, 4], idx: 3
       -Num: [:e, 5], idx: 4
       +Num: , idx: 0
       +Num: , idx: 1
       +Num: , idx: 2
       +Num: , idx: 3
       +Num: , idx: 4
       
     Shared Example Group: "#each_with_index" called from ./spec/enumerables_spec.rb:112
     # ./spec/enumerables_spec.rb:77:in `block (4 levels) in <top (required)>'
```

This means that it's unable to handle hashes in it's input.

Please check that it can produce the desired output when used thus:

```
        enum = { a: 1, b: 2, c: 3, d: 4, e: 5 }
        my_each_with_index_output = ''
        block = proc { |num, idx| my_each_with_index_output += "Num: #{num}, idx: #{idx}\n" }
        enum.each_with_index(&block)
        each_with_index_output = my_each_with_index_output.dup
        my_each_with_index_output = ''
        enum.my_each_with_index(&block)

# Both my_each_with_index_output and each_with_index_output should contain the same value:

        p my_each_with_index_output
        p each_with_index_output
```

Reply:
my_each_with_index_output 
=> "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
each_with_index_output                                           
=> "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"